[TG-1419] Specialization of generics

majakusber opened the pull request:
This re-implements the specialization of generics - instead of creating new symbols in the symbol table for a specialized generic type, we usep a map of pairs: generic parameter -> stack of types, to keep track of the concrete type of parameters in each context (and context depth). Moreover, the implementation also works with generic superclasses and interfaces, which was not the case in the old version.

Commit:
Introduce a map of parameter-type pairs

The map of generic parameters and a stack of their concrete types in the
various depths of the current scope.

Commit:
Updating and extending utility functions

mgudemann commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -322,8 +322,13 @@ void ci_lazy_methodst::initialize_all_needed_classes_from_pointer(
+  // we pass an empty map to convert pointer type as we do not need to
+  // record the mapping of generic parameters to concrete types in the
+  // current context at this point (we are not initializing objects, just
+  // loading necessary classes). all the types that we need to load for
```
typo `. all` -> `. All`

mgudemann commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -68,6 +68,17 @@ class java_object_factoryt
+  /// Note that in different depths of the scope the parameters can be
+  /// specialized with different types so we keep a stack of types
+  /// for each parameter.
+  std::unordered_map<irep_idt, std::stack<reference_typet>, irep_id_hash>
```
please use a `typedef` for this

mgudemann commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +509,79 @@ class recursion_set_entryt
+  /// controlled map.
+  ~generic_recursion_map_keyst()
+  {
+    for(const auto key : erase_keys)
```
I would prefer a `std::remove_if` here

majakusber replied:
I'm not quite sure how to use it here. For each `key` we just pop from its stack, but don't remove the key itself from the map (unless the stack becomes empty).

mgudemann commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +509,79 @@ class recursion_set_entryt
+    }
+  }
+
+  generic_recursion_map_keyst(const generic_recursion_map_keyst &) = delete;
```
would you mind adding a comment here to explain what this does?

mgudemann commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +509,79 @@ class recursion_set_entryt
+    const std::vector<reference_typet> &types)
+  {
+    INVARIANT(erase_keys.empty(), "insert_pairs should only be called once");
+    PRECONDITION((parameters.size() == types.size()) && parameters.size() > 0);
```
`!parameters.empty() && ( ..size() == ...size())`

mgudemann commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +509,79 @@ class recursion_set_entryt
+  {
+    INVARIANT(erase_keys.empty(), "insert_pairs should only be called once");
+    PRECONDITION((parameters.size() == types.size()) && parameters.size() > 0);
+    for(size_t i = 0; i < parameters.size(); i++)
```
why not ranged for over `parameters` ?

majakusber replied:
I need the index of the current parameter to retrieve the type at the same index from `types`.

mgudemann commented on src/java_bytecode/java_types.h:
```diff
@@ -209,6 +213,19 @@ class java_generic_typet:public reference_typet
+  void replace(const typet &argument, const typet &type)
+  {
+    PRECONDITION(is_reference(type));
+    for(auto &arg : generic_type_arguments())
```
`std::replace_if` ?

mgudemann commented on src/java_bytecode/java_types.h:
```diff
@@ -393,6 +410,46 @@ to_java_implicitly_generic_class_type(java_class_typet &type)
+/// generic class type, false otherwise
+inline const bool is_java_any_generic_class_type(const typet &type)
+{
+  const bool value = is_java_generic_class_type(type) ||
```
the intermediate variable isn't needed here

mgudemann reviewed (changes requested):
one question, why are some of the unit test class / java files deleted ?

majakusber commented:
@mgudemann I deleted the folder `unit/java_bytecode/generate_concrete_generic_type` which contained unit tests for checking whether a new symbol was created in the symbol table for specialized generic classes with correct types. Now we do not create such a symbol, the specialization is now done on-the-fly. I also removed some utility functions used in those tests only.

Commit:
Adding and updating unit tests

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -68,6 +68,17 @@ class java_object_factoryt
 
+  /// Every time the non-det generator visits a type and the type is generic
+  /// (either a struct or a pointer), all its generic parameters must have
+  /// a concrete type at this point (otherwise an object cannot be created).
```
This comment doesn't seem quite correct (it isn't the case that they must since the function might be generic)

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +509,79 @@ class recursion_set_entryt
+      (*generic_recursion_map.find(key)).second.pop();
+      if((*generic_recursion_map.find(key)).second.empty())
+      {
+        generic_recursion_map.erase(key);
```
Does it not need to restore the old value?

majakusber replied:
We pop from the stack once (three lines above), so the old value is now the top value. We only remove the whole entry for the key if the stack is empty.

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -393,6 +397,46 @@ to_java_implicitly_generic_class_type(java_class_typet &type)
+/// \param type Source type
+/// \return The vector of implicitly generic and generic type parameters of
+/// the given type.
+inline const std::vector<java_generic_parametert> get_all_generic_parameters
```
Why are these methods marked as `inline`?

majakusber replied:
The functions are reasonably small (at least the first one) and may end up being called quite often, I thought. Would you remove the `inline`?

thk123 replied:
In C++ inline no longer means inline this function. It is, at best, a hint to the compiler that this might be a good candidate to inline. 

Modern compilers will ignore this as they (almost) always know better and will inline according to their own criteria (including functions that are not marked inline). Instead, it is mainly for AFAIK allowing [more than one definition](http://en.cppreference.com/w/cpp/language/inline).

majakusber replied:
I see, thanks for the link. I will remove the `inline` then.

majakusber replied:
Seems that to keep the functions in the header (which I think makes sense, since similar functions for java types are in the header too) I have to keep the `inline`. Otherwise I get errors about multiple definitions.

thk123 replied:
Since non-trivial I would favour in the cpp file, but just preference

majakusber replied:
Since I only used this function while updating the generic map and I pulled that map into a separate file, I moved the function to that file and it's not inlined anymore.

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -393,6 +397,46 @@ to_java_implicitly_generic_class_type(java_class_typet &type)
+/// \param type The type to check
+/// \return True if the given type is a generic class type or an implicitly
+/// generic class type, false otherwise
+inline const bool is_java_any_generic_class_type(const typet &type)
```
I'm not sure this method improves readability - I wouldn't know what "any generic class" means so would have to look it up, probably would prefer to just use the condition. Or better yet if we can come up with a better name (I'll have a think)

thk123 replied:
`is_class_with_generic_parameter`?

majakusber replied:
One option would be to have `java_explicitly_generic_class_type` and `java_implicitly_generic_class_type` that would both extend `java_generic_class_type` (there is already a class like this but it's meaning now is only explicitly generic, not implicitly). Then we could use the word generic everywhere in the code with the expected meaning, only the lower level code would distinguish between the two.

thk123 replied:
That sounds reasonable - is it a big change? If so, would it be reasonable to do a separate PR?

majakusber replied:
I think it would make a good sized PR and the issue is independent of this ticket. I can create a Jira ticket for it.

majakusber replied:
https://diffblue.atlassian.net/browse/TG-2639

thk123 replied:
I think in the mean time (since we don't know when we're going to get to this TD) I'd rename this function (or eliminate it if you prefer) according to above suggestions

majakusber replied:
I will remove it for now.

thk123 commented on src/java_bytecode/ci_lazy_methods.cpp:
```diff
@@ -322,8 +322,13 @@ void ci_lazy_methodst::initialize_all_needed_classes_from_pointer(
+  // current context at this point (we are not initializing objects, just
+  // loading necessary classes). All the types that we need to load for
+  // future specialization must anyway appear on left-hand-side somewhere.
+  const pointer_typet &subbed_pointer_type =
```
Note to check there is a test verifying this assumption

majakusber replied:
I'm not sure how to check that the assumption is always true.. any idea what the test should have in it? I was trying to think of an example, something like this:
```
class A<T>
{
  public T a;
}

class B extends A<Integer>
{
}

public class base_test
{
  public B f;
  public A g = new A<Boolean>();

  public Integer test()
  {
    if(f.a==null && g.a)
    {
      return 0;
    }
    else if(f.a!=null)
    {
      return f.a;
    }
    else
    {
      return 1;
    }
  }
}
```
But this does not compile with an error:
```
error: bad operand types for binary operator '&&'      if(f.a==null && g.a)
```
Which supports the assumption.

thk123 replied:
I guess I was just thinking:
```
public class base_test {
  public static String test(A<Integer> input) {
    if(input != null)
      return input.toString();
    else
      return "";
}
```
Then this should lazy load `Integer.toString()` 

(even better if it includes another test which is the same except the `input` type is `B` .

majakusber replied:
I tested it, with `A<Integer> input` it loads the correct method, but not with `B input`. Will come up with a solution...

thk123 replied:
I'm happy for this to be spun out into a separate ticket (with appropriate notes for QA) as this is still a big step forward, unless it seems super simple. (Though of course adding a test demonstrating the problem would be appreciated)

majakusber replied:
It turns out it's not super simple :) I pulled the issue out into a separate ticket https://diffblue.atlassian.net/browse/TG-2664. The ticket contains the example from above.

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -213,6 +213,20 @@ class java_generic_typet:public reference_typet
       add(ID_type_variables).get_sub());
   }
+
+  /// Replace every occurrence of an argument with a given type.
```
generic argument? As opposed to argument to method call?

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -213,6 +213,20 @@ class java_generic_typet:public reference_typet
+    PRECONDITION(is_reference(type));
+    auto isArgument = [&](reference_typet t) { return (t == argument); };
+    std::replace_if(
+      generic_type_arguments().begin(),
```
Where has this come from? `generic_type_arguments` is not an input to this method, but this method doesn't belong to  a class?

majakusber replied:
This whole function `replace` is in class `java_generic_typet`. The `generic_type_arguments()` function is in the same class.

thk123 replied:
Ah I know what confused me - method body inside the class definition, would prefer since not completely trivial to move into the source file.

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -213,6 +213,20 @@ class java_generic_typet:public reference_typet
+  void replace(const typet &argument, const typet &type)
+  {
+    PRECONDITION(is_reference(type));
+    auto isArgument = [&](reference_typet t) { return (t == argument); };
```
`is_argument` (though name sounds like a bool rather than a predicate, perhaps `check_is_argument`?

thk123 commented on src/java_bytecode/select_pointer_type.cpp:
```diff
@@ -11,15 +11,101 @@
+  const irep_idt &parameter_name =
+    to_java_generic_parameter(parameter).get_name();
+
+  if(!generic_recursion_map.count(parameter_name))
```
Implict conversion from `int` to `bool` should be avoided IMO. I'd prefer either:

```C++
generic_recursion_map.count(parameter_name) != 0
```
or
```C++
generic_recursion_map.find(parameter_name) != generic_recursion_map.end()
```

thk123 commented on src/java_bytecode/select_pointer_type.cpp:
```diff
@@ -11,15 +11,101 @@
+    return find_concrete_type(type, generic_recursion_map);
+  }
+  else if(is_java_generic_type(type))
+  {
```
I don't follow this block - going to come back to it when I've reviewed the test

thk123 replied:
So if I've understood correctly: this is required because if we have:
```java
class Gen<T> { T field; }
class Gen2<U> { Gen<U> genField; }
Gen2<Integer> param;
```
We need to make sure we map Gen::T -> Gen2::U. `Gen2::U`  is mapped `Integer` in the `generic_recursion_map`. 

What I don't understand is why can't this first relation be in the map also? 

Long term thoughts: when we deal with generic functions, we might have for example that Gen2::U is one of `{Object, Integer, String}`, then here we'd need duplicate however we handle having multiple types. 

Let's discuss this offline tomorrow

majakusber replied:
This is the tree of calls that happen for your example in the code (hope it's not more confusing than helping):
call `gen_nondet_init` on pointer to `Gen2<Integer>` (this adds `Gen2::U -> Integer` to the map)
- call `gen_nondet_pointer_init` on `Gen2<Integer>`
- - call `specialize_generics` on `Gen2<Integer>` (this is java generic type - else if branch)
- - - call `specialize_generics` on `Integer` (this is not parameter, not generic type - else branch)
- - call `gen_nondet_init` on class `Gen2`
- - - call `gen_nondet_struct_init` on `Gen2`
- - - - call `gen_nondet_init` on pointer to `Gen<U>` as a field of `Gen2` (this adds `Gen::T -> Gen2::U` to the map)
- - - - - call `gen_nondet_pointer_init` on `Gen<U>`
- - - - - - call `specialize_generics` on `Gen<U>` (this is java generic type - else if branch)
- - - - - - - call `specialize_generics` on `U` (this is parameter - if branch) which returns `Integer`
- - - - - - call `gen_nondet_init` on class `Gen`
- - - - - - - call `gen_nondet_struct_init` on `Gen`
- - - - - - - - call `gen_nondet_init` on pointer to `T` as field of `Gen`
- - - - - - - - - call `gen_nondet_pointer_init` on `T`
- - - - - - - - - - call `specialize_generics` on `T` (this is parameter - if branch) which returns `Integer`

So you do have all you need in the map. In `specialize_generics`, you do not manipulate the map, just recursively go over the generic type (might have multiple and/or nested generic types and parameters in it such as `Gen<GenTwo<U, Integer>>`).

thk123 replied:
That is helpful, but I don't see how the result is used anywhere? Seems like the first call adds `Get2::U -> Integer` to the map, so we don't care when we read it on 

>               - call specialize_generics on U (this is parameter - if branch) which returns Integer

majakusber replied:
After an offline discussion, I believe this indeed can be reduced to only specialize generics if the type at hand is a generic parameter.

thk123 commented on src/java_bytecode/select_pointer_type.h:
```diff
@@ -21,7 +22,22 @@ class select_pointer_typet
   virtual pointer_typet convert_pointer_type(
-    const pointer_typet &pointer_type, const namespacet &ns) const;
+    const pointer_typet &pointer_type,
+    const std::unordered_map<irep_idt,
```
Probably worth `typedef`ing this type

thk123 commented on src/java_bytecode/generate_java_generic_type.cpp:
```diff
@@ -1,343 +0,0 @@
-/// Small auxiliary function, to print a single generic argument name.
-/// \param argument argument type
-/// \return printed name of the argument
-static std::string argument_name_printer(const reference_typet &argument)
```
Probably worth taking a note of this commit somewhere, as some of the utilities seem like they will be useful?

majakusber replied:
I agree, I made a note about this function in my notes. I also added a comment in TG-1278 and TG-1871 which seem to be the most relevant open tickets for pretty printing generic types.

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -55,18 +62,59 @@ require_goto_statements::find_struct_component_assignments(
+              to_member_expr(superclass_compound);
+
+            if(
+              superclass_member_expr.get_component_name() ==
```
This condition is hard to follow (partly down to line breaks) consider breaking out the two conditions into intermediate variables

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -55,18 +62,59 @@ require_goto_statements::find_struct_component_assignments(
-        if(
-          symbol.get_identifier() == structure_name &&
-          member_expr.get_component_name() == component_name)
+        if(superclass_name.has_value())
```
I found this a bit hard to follow - might it be simpler to construct the expected expr tree and then check if they are equal?

majakusber replied:
I only pass `irep_id`s to this function and to construct the expr, I would need to full types. As mentioned in a comment above, I tried to make things more readable here, let me know if you find it sufficient.

thk123 reviewed (commented):
Got to commit Adding and updating unit tests so appologies if comments are addressed in later commits

thk123 commented on unit/goto-programs/goto_program_generics/generic_bases_test.cpp:
```diff
@@ -0,0 +1,458 @@
+
+    WHEN("The entry point function is generated")
+    {
+      const auto &entry_point_function =
```
In the interest of getting these tests to be as short as possible, I wonder if you could pull out (into this file) hyper specific utilities to:

- get the `code_blockt` for the entry function
- get the this pointer assignment (with the checks that there is precisely 1 non-null assignment)

thk123 replied:
And maybe even getting the rhs of that assignment since that is all we care about

thk123 commented on unit/goto-programs/goto_program_generics/generic_parameters_test.cpp:
```diff
@@ -0,0 +1,476 @@
+
+      THEN(
+        "Object 'this' has field 'field_input' of type "
+          "SimpleWrapper<IWrapper>")
```
Comment needs updating (just of type SimpleWrapper)

thk123 replied:
And elsewhere in this file I guess

thk123 reviewed (commented):
Nothing blocking on the unit tests, just some suggestions for reducing duplication

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -61,24 +61,28 @@ require_goto_statements::find_struct_component_assignments(
           if(
             member_expr.get_component_name() == component_name &&
-            superclass_compound.id() == ID_member)
+            member_expr.op0().id() == ID_member)
```
`compound()` rather than `op0`

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -61,24 +61,28 @@ require_goto_statements::find_struct_component_assignments(
-              superclass_member_expr.symbol().get_identifier() ==
-                structure_name)
+              superclass_expr.get_component_name() == supercomponent_name &&
+              superclass_expr.symbol().get_identifier() == structure_name)
```
Would still rather if we could de-duplicate this if block with the existing one (i.e. if super class check the super class component then have one bit of code that checks the final component), but it is now much clearer so non-blocking.

thk123 commented on unit/testing-utils/require_goto_statements.cpp:
```diff
@@ -33,6 +34,18 @@ require_goto_statements::get_all_statements(const exprt &function_value)
+  // Retrieve __CPROVER_start
+  const symbolt *entry_point_function =
+    symbol_table.lookup(goto_functionst::entry_point());
+
```
To avoid crashing probably worth putting a `REQUIRE(entry_point_function != nullptr)` in here

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +508,79 @@ class recursion_set_entryt
+    const std::vector<reference_typet> &types)
+  {
+    INVARIANT(erase_keys.empty(), "insert_pairs should only be called once");
+    PRECONDITION(!parameters.empty() && (parameters.size() == types.size()));
```
Missed this earlier, but maybe a `std::vector<std::pair<java_generic_parametert, reference_typet>>` (or even `std::mapr<java_generic_parametert, reference_typet>`) would encode this precondition at compile time.

thk123 replied:
And would simplify the below for loop as you could just use a range based for loop over the pairs

thk123 replied:
Though I see this makes calling this more complicated - up to you what you do here

majakusber replied:
Yes, I would have to construct the vector of pairs before calling the function and this precondition would have to appear there somewhere anyway. I can create a separate function for the construction and call it at the beginning of this function.

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +508,79 @@ class recursion_set_entryt
+    {
+      const irep_idt &key = parameters.at(i).get_name();
+      const reference_typet &value = types.at(i);
+      // Only add the pair if the type is not the parameter itself. This can
```
Consider adding an example to this comment as I'm struggling to figure out what it means

thk123 replied:
This is mapping `A::T` -> `A::T` right?

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +508,79 @@ class recursion_set_entryt
+        !(is_java_generic_parameter(value) &&
+          to_java_generic_parameter(value).get_name() == key))
+      {
+        if(!generic_recursion_map.count(key))
```
Implict conversion from int to bool, replace with `== 0`

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +508,79 @@ class recursion_set_entryt
+        (*generic_recursion_map.find(key)).second.push(value);
+
+        // We added something, so pop it when this is destroyed:
+        erase_keys.push_back(parameters.at(i).get_name());
```
`parameters.at(i).get_name()` = `key` no?

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -498,6 +508,79 @@ class recursion_set_entryt
+  {
+    for(const auto key : erase_keys)
+    {
+      PRECONDITION(generic_recursion_map.count(key));
```
Implicit conversion from int to bool

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -209,6 +213,20 @@ class java_generic_typet:public reference_typet
+  /// Replace every occurrence of a generic type argument with a given type.
+  /// \param argument generic type argument that should be replaced
+  /// \param type new type that the argument should be replaced with
+  void replace(const typet &argument, const typet &type)
```
Outstanding without reply:
> Ah I know what confused me - method body inside the class definition, would prefer since not completely trivial to move into the source file.

Preference/non-blocking

thk123 reviewed (commented):
Nothing blocking - some final suggestions (all review comments from previous reviews either addressed or re-mentioned in this review)

thk123 commented on src/java_bytecode/generic_recursion_map_keys.cpp:
```diff
@@ -0,0 +1,158 @@
+/// \return The vector of implicitly generic and (explicitly) generic type
+/// parameters of the given type.
+const std::vector<java_generic_parametert> get_all_generic_parameters
+  (const typet &type)
```
This formatting looks suspect?

thk123 commented on src/java_bytecode/generic_recursion_map_keys.cpp:
```diff
@@ -0,0 +1,158 @@
+  const std::vector<java_generic_parametert> &parameters,
+  const std::vector<reference_typet> &types)
+{
+  PRECONDITION(!parameters.empty() && (parameters.size() == types.size()));
```
While I agree parameters being empty seems odd, not really a pre-req for this function (in can pair up two empty lists into an empty list of pairs) might not be useful now but in some circumstances might make using this utility more complicated. 

You might consider pulling this into a totally templated function in util to pair arbitary vectors in to a vector of pairs,
 but no need to do this as part of this PR.

thk123 commented on src/java_bytecode/generic_recursion_map_keys.cpp:
```diff
@@ -0,0 +1,158 @@
+{
+  PRECONDITION(!parameters.empty() && (parameters.size() == types.size()));
+  std::vector<std::pair<java_generic_parametert, reference_typet>> vector;
+  for(size_t i = 0; i < parameters.size(); i++)
```
This [SO answer](https://stackoverflow.com/questions/18478385/convert-two-vectors-of-int-with-the-same-length-into-one-vector-of-pairs-of-int?lq=1) has a way of doing it using std::transform, though I'm not sure it is better.

thk123 replied:
For consistency, prefer `++i` in a for loop (or use iterators)

thk123 commented on src/java_bytecode/generic_recursion_map_keys.cpp:
```diff
@@ -0,0 +1,158 @@
+  const std::vector<reference_typet> &types)
+{
+  INVARIANT(erase_keys.empty(), "insert_pairs should only be called once");
+  const std::vector<std::pair<java_generic_parametert, reference_typet>>
```
I'd probably typedef `std::vector<std::pair<java_generic_parametert, reference_typet>>` somewhere

majakusber replied:
I removed the function `pair_up` and the construction is done within `insert_pairs` so this type is explicitly used only once now.

thk123 commented on src/java_bytecode/generic_recursion_map_keys.h:
```diff
@@ -0,0 +1,60 @@
+    const std::vector<java_generic_parametert> &parameters,
+    const std::vector<reference_typet> &types);
+
+public:
```
Nit: I think it is helpful to keep the public interface at the top of the class

thk123 commented on src/java_bytecode/generic_recursion_map_keys.h:
```diff
@@ -0,0 +1,60 @@
+class generic_recursion_map_keyst
+{
+  /// Recursion map to modify
+  generic_recursion_mapt &generic_recursion_map;
```
Nit: prefer keeping private members at the bottom so easy to see the interface

thk123 commented on src/java_bytecode/generic_recursion_map_keys.cpp:
```diff
@@ -0,0 +1,158 @@
+  }
+}
+
+///
```
Either document or don't :stuck_out_tongue:

thk123 commented on src/java_bytecode/generic_recursion_map_keys.cpp:
```diff
@@ -0,0 +1,158 @@
+}
+
+///
+/// \param pointer_type
```
docs missing

mgudemann commented on src/java_bytecode/java_types.h:
```diff
@@ -120,10 +120,14 @@ class java_generic_parametert:public reference_typet
   {
-    const std::string &parameter_name =
-      type_variable().get_identifier().c_str();
+    const std::string &parameter_name = get_name().c_str();
```
why not `id2string` instead of `c_str` here?

majakusber replied:
This function is actually not used anymore (was part of the old implementation for specialization) so will remove.

thk123 replied:
This still seems to be here? Perhaps removed in a later commit

mgudemann commented on unit/goto-programs/goto_program_generics/GenericFields.java:
```diff
@@ -0,0 +1,99 @@
+// THIS FILE HAS A COPY IN test-gen/unit/test-java-gen
```
does git allow for symbolic links? we don't have to support Windows in TG

majakusber replied:
I removed this because the tests in test-gen will be changed to not be just a copy.

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -68,6 +69,16 @@ class java_object_factoryt
+  /// from the map when the scope changes. Note that in different depths
+  /// of the scope the parameters can be specialized with different types
+  /// so we keep a stack of types for each parameter.
+  generic_recursion_mapt generic_recursion_map;
```
Is `recursion` actually the right name? Sorry should have spotted earlier, but it is more of a `generic_type_specialisation_map` no?

thk123 commented on src/java_bytecode/java_types.cpp:
```diff
@@ -802,3 +802,18 @@ void get_dependencies_from_generic_parameters(
+/// with a given type.
+/// \param argument generic type argument that should be replaced
+/// \param type new type that the argument should be replaced with
+void java_generic_typet::replace(const typet &argument, const typet &type)
```
This method is no longer used I think?

thk123 reviewed (commented):
All previous comments have been addressed, commits look reasonable. This needs for sure the TG pointer bump updating. I also think it is worth reconsidering the name of the generic_recursion_map as I'm not really sure what the recursive element is.

mgudemann reviewed (approved):
LGTM, two small comments but not blocking of course

Commit:
Use the map of parameter-type pairs to specialize generics

Commit:
Remove the old way of specializing generics

We now specialize generics on-the-fly using a context-sensitive map of
parameter-type pairs

thk123 commented on src/java_bytecode/generic_parameter_specialization_map_keys.h:
```diff
@@ -0,0 +1,69 @@
+    const typet &symbol_struct);
+
+private:
+  /// Recursion map to modify
```
Generic parameter map to modify

thk123 commented on src/java_bytecode/generic_parameter_specialization_map_keys.cpp:
```diff
@@ -0,0 +1,152 @@
+  }
+}
+
+/// Add pairs of parameters and their types for a generic symbol type to the
```
Perhaps improve this comment by explaining that the pairs are for each generic parameter and why you would use this overload of the method vs. the above one (e.g. something about being used for the base class?)

thk123 commented on src/java_bytecode/generic_parameter_specialization_map_keys.cpp:
```diff
@@ -0,0 +1,152 @@
+  }
+}
+
+/// Add pairs of parameters and their types for a generic pointer type to the
```
Perhaps improve this comment by explaining that the pairs are for each generic parameter

thk123 reviewed (approved):
Some suggestions for comments but no need for re-review

thk123 commented:
@marek-trtik This is the PR you will be interested in when it is merged (tagging you so you'll get a notification when merged)

marek-trtik commented:
@thk123 Thanks a lot!

majakusber commented:
@thk123 Can you please have one last look? I made some changes to the commit 'Use the map..' so that a replacement pointer type is now created for array types that use generic parameters.

thk123 commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -753,12 +753,21 @@ void java_object_factoryt::gen_nondet_pointer_init(
+    (is_reference(replacement_pointer_type) && is_reference(pointer_type) &&
+     to_reference_type(replacement_pointer_type)
+       .
+       operator!=(to_reference_type(pointer_type))))
```
Don't use `.operator!=`, just use `!=`?
Also I'm not sure how this works. Perhaps for readability it could be broken into two local booleans. None the less, I can't see how `r == p` but casting it using `to_reference_type` and then checking again can produce a different result?

majakusber replied:
I replaced the operator with a function called `equal_java_types` and put it in `java_types.cpp`

thk123 reviewed (commented):
Will have a detailed look tomorrow - most important question, do you have a test that now passes that previously would have failed?

Commit:
Fix a bug in function for extracting generic interface reference

thk123 commented on src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -151,7 +151,16 @@ static optionalt<std::string> extract_generic_interface_reference(
 
-    start = signature.value().find("L" + interface_name + "<", start);
+    // if the interface name includes package name, convert dots to slashes
+    std::string interface_name_slash_to_dot = interface_name;
```
When doing this kind of string manipulation, it is probably worth also checking that inner classes work (since the whole `$`/`.` fiasco) 

For example (though I guess with some generics in there as well). 
```
class Outer {
  public interface InnerInterface { }
}
class TestClass implements Outer.InnerInterface {}
```

majakusber replied:
I tried the example, this is not causing a problem since both in class signatures and in our representation we use `$` as separator for inner classes.

thk123 reviewed (commented):
A unit test with an interface in a package and a unit test with an inner class interface would be good - lets talk tomorrow if this is a sensible thing to spin out or whether should be tacked on.
